Neos.Media/Classes/Domain/Model/AssetSource/Neos/NeosAssetSource.php

```diff
@@ -0,0 +1,184 @@
+<?php
+namespace Neos\Media\Domain\Model\AssetSource\Neos;
+
+/*
+ * This file is part of the Neos.Media package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+  *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\Flow\Core\Bootstrap;
+use Neos\Flow\Http\HttpRequestHandlerInterface;
+use Neos\Flow\Http\Uri;
+use Neos\Flow\Mvc\ActionRequest;
+use Neos\Media\Domain\Model\AssetInterface;
+use Neos\Media\Domain\Model\AssetSource\AssetProxyRepository;
+use Neos\Media\Domain\Model\AssetSource\AssetSource;
+use Neos\Media\Domain\Service\AssetService;
+use Neos\Media\Domain\Service\ThumbnailService;
+use Neos\Flow\Annotations as Flow;
+use Neos\Media\Exception\AssetServiceException;
+use Neos\Media\Exception\ThumbnailServiceException;
+use Psr\Http\Message\UriInterface;
+
+/**
+ * Asset source for Neos native assets
+ */
+final class NeosAssetSource implements AssetSource
+{
+    /**
+     * @var string
+     */
+    private $assetSourceIdentifier;
+
+    /**
+     * @var NeosAssetProxyRepository
+     */
+    private $assetProxyRepository;
+
+    /**
+     * @Flow\Inject
+     * @var ThumbnailService
+     */
+    protected $thumbnailService;
+
+    /**
+     * @Flow\Inject
+     * @var AssetService
+     */
+    protected $assetService;
+
+    /**
+     * @Flow\Inject
+     * @var Bootstrap
+     */
+    protected $bootstrap;
+
+    /**
+     * @Flow\InjectConfiguration(path="asyncThumbnails")
+     * @var bool
+     */
+    protected $asyncThumbnails;
+
+    /**
+     * @param string $assetSourceIdentifier
+     * @param array $assetSourceOptions
+     */
+    public function __construct(string $assetSourceIdentifier, array $assetSourceOptions)
+    {
+        if (preg_match('/^[a-z][a-z0-9-]{0,62}[a-z]$/', $assetSourceIdentifier) !== 1) {
+            throw new \InvalidArgumentException(sprintf('Invalid asset source identifier "%s". The identifier must match /^[a-z][a-z0-9-]{0,62}[a-z]$/', $assetSourceIdentifier), 1513329665386);
+        }
+        $this->assetSourceIdentifier = $assetSourceIdentifier;
+
+        foreach ($assetSourceOptions as $optionName => $optionValue) {
+            switch ($optionName) {
+                case 'asyncThumbnails':
+                    // If the option value is empty, preserve the default value injected from the Neos:Media:asyncThumbnails setting:
+                    if (!empty($optionValue)) {
+                        if (!is_bool($optionValue)) {
+                            throw new \InvalidArgumentException(sprintf('Asset source option "%s" specified for Neos asset source "%s" must be either true or false. Please check your settings.', $optionName, $assetSourceIdentifier), 1522927471208);
+                        }
+                        $this->asyncThumbnails = (bool)$optionValue;
```

Comment:
If we really need to be so strict as to chedck for boolean above, we don't really need to cast here...